src/sql/services/notebook/notebookServiceImpl.ts

```diff
@@ -169,14 +181,25 @@ export class NotebookService extends Disposable implements INotebookService {
 			throw new Error(localize('notebookUriNotDefined', 'No URI was passed when creating a notebook manager'));
 		}
 		let uriString = uri.toString();
-		let manager = this._managers.get(uriString);
-		if (!manager) {
-			manager = await this.doWithProvider(providerId, (provider) => provider.getNotebookManager(uri));
-			if (manager) {
-				this._managers.set(uriString, manager);
+		let manager: INotebookManager[] = this._managers.get(uriString);
+		// If manager already exists for a given notebook, return it
+		if (manager) {
+			let index = manager.findIndex(m => m.providerId === providerId);
+			if (index && index >= 0) {
+				return manager[index];
 			}
 		}
-		return manager;
+		let newManager = await this.doWithProvider(providerId, (provider) => provider.getNotebookManager(uri));
+
+		// If no managers exist for a notebook, add this manager
+		// Otherwise, add manager to array
+		if (!manager) {
+			this._managers.set(uriString, [newManager]);
```

Comment:
OK, can we also rename `this._managers` to `this._managersMap` as it's confusing to have this + managers array + manager instance.